Allow selecting image quality among multiple images

Stypox opened the pull request:
#### What is it?
- [x] Feature (user facing)

#### Description of the changes in your PR
- This PR adapts to https://github.com/TeamNewPipe/NewPipeExtractor/pull/889, which introduces breaking changes regarding images. Instead of providing just one URL per image (e.g. `getThumbnailUrl()`), **the extractor now provides a `List<Image>`** (e.g. `getThumbnails()`), and each image has `width`, `height` and `resolutionLevel` attributes that allow determining an image's quality.
- This PR introduces a preference that allows users to either disable images completely, or to **choose a level amongst "low", "medium" and "high"**. This setting replaces the "Load thumbnails" switch.
- The strategy that determines which image is chosen based on the user preference is [here](https://github.com/Stypox/NewPipe/blob/multiple-images/app/src/main/java/org/schabi/newpipe/util/image/ImageStrategy.java#L54), along with documentation.
- **In the database only one image URL is still stored, instead of the whole `List<Image>`.** The saved URL is the preferred one at the time of saving. This keeps it simple, avoids needing to do database migrations (not really a big deal, but still) and saves storage. Otherwise we would probably double the database size just to store a whole lot of images that will never be all used anyway. The only downside is that upon changing preferred resolution level in settings, only newly fetched `List<Image>`s will have the correct level, while videos in history/feed will not. I don't think it's such a big deal, also because when opening a (remote) video (or playlist), the thumbnail(s) would get updated to the newly set preferred resolution level.
- The description tab in the video detail fragment now shows all available thumbnails, highlighting the one being chosen and allowing to tap on them to open in browser.

#### Before/After Screenshots/Screen Record

<table>
  <tr>
    <th>Before preference</th>
    <th>Before list</th>
  </tr>
  <tr>
    <td><img src="https://user-images.githubusercontent.com/36421898/235771294-122f0252-8174-4795-9659-d5013b8bd6a4.png" width="200px"/></td>
    <td><img src="https://user-images.githubusercontent.com/36421898/235771165-b2e9485b-d1e2-4c77-83c7-9e1a3089943a.png" width="200px"/></td>
  </tr>
</table>

<table>
  <tr>
    <th>Preference</th>
    <th>Low (list)</th>
    <th>Medium (list)</th>
    <th>High (video details)</th>
    <th>Description</th>
  </tr>
  <tr>
    <td><img src="https://user-images.githubusercontent.com/36421898/235770044-611c2ca6-17a8-486f-9c7a-8b8b0801b81a.png" width="200px"/></td>
    <td><img src="https://user-images.githubusercontent.com/36421898/235770127-c213157a-018f-45fc-9913-04039410bc90.png" width="200px"/></td>
    <td><img src="https://user-images.githubusercontent.com/36421898/235770183-e46e8a97-34af-4e4e-a3ca-e8b661213dd1.png" width="200px"/></td>
    <td><img src="https://user-images.githubusercontent.com/36421898/235772899-8445e751-f529-4752-8f56-72d8bc7c2adf.png" width="200px"/></td>
    <td><img src="https://user-images.githubusercontent.com/36421898/235770481-22ae8c21-d0f5-401c-8bbc-0c7dcfe690d8.png" width="200px"/></td>
  </tr>
</table>

#### TODO
- [x] Add description to PR
- [x] Allow user to choose thumbnail quality
- [x] Refine preferred image choosing method
- [x] Decide whether it's fine to save only one thumbnail in the database - Yes, it's fine.
- [x] How should `PlaylistRemoteEntity.isIdenticalTo` behave? - it's fine if the thumbnail is updated in the database if the current preference has changed (since that implies `choosePreferredImage` returns a different url than the one saved in the database, and hence `isIdenticalTo` returns `false`)
- [x] The description fragment should show all thumbnail URLs (and maybe highlight the preferred one being chosen)
- [x] Add comments to the code, and possibly rename some variables
- [x] Add tests for `ImageStrategy`
- [ ] Update extractor commit after channel tabs PR is merged

#### Fixes the following issue(s)
- Fixes #3121
- Closes #3190 because we chose a different approach
- Closes #1782
- Closes #7656
- Also implements the ability to close #10055, by sending the highest resolution image URL to an external app such as a browser

#### Relies on the following changes

https://github.com/TeamNewPipe/NewPipeExtractor/pull/889

#### APK testing
The APK can be found by going to the "Checks" tab below the title. On the left pane, click on "CI", scroll down to "artifacts" and click "app" to download the zip file which contains the debug APK of this PR.

At the moment the highest quality thumbnails will load everywhere, and that's the only noticeable difference with normal NewPipe, until a quality setting is added.

#### Due diligence
- [x] I read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md).

Jean-BaptisteC commented:
Crash when i try to refresh feed
```
-------------------------------------
java.lang.NullPointerException: Parameter specified as non-null is null: method org.schabi.newpipe.local.feed.service.FeedLoadService$onStartCommand$2.accept, parameter error
	at org.schabi.newpipe.local.feed.service.FeedLoadService$onStartCommand$2.accept(Unknown Source:7)
	at org.schabi.newpipe.local.feed.service.FeedLoadService$onStartCommand$2.accept(FeedLoadService.kt:98)
	at io.reactivex.rxjava3.internal.observers.BiConsumerSingleObserver.onSuccess(BiConsumerSingleObserver.java:56)
	at io.reactivex.rxjava3.internal.operators.single.SingleDoOnSubscribe$DoOnSubscribeSingleObserver.onSuccess(SingleDoOnSubscribe.java:77)
	at io.reactivex.rxjava3.internal.operators.single.SingleObserveOn$ObserveOnSingleObserver.run(SingleObserveOn.java:81)
	at io.reactivex.rxjava3.android.schedulers.HandlerScheduler$ScheduledRunnable.run(HandlerScheduler.java:123)
	at android.os.Handler.handleCallback(Handler.java:942)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loopOnce(Looper.java:201)
	at android.os.Looper.loop(Looper.java:288)
	at android.app.ActivityThread.main(ActivityThread.java:7884)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936)
-------------------------------------
```

Stypox commented:
That's unrelated to this PR and is fixed in the channel tabs PR. I don't know why suddenly that issue started arising, as the code there didn't change... Probably a kotlin update

Isira-Seneviratne commented on app/src/main/java/org/schabi/newpipe/util/image/ImageStrategy.java:
```diff
@@ -0,0 +1,203 @@
+                })
+                // then each level's group is further split into two subgroups, one with known image
+                // size (which is also the preferred subgroup) and the other without
+                .thenComparing((a, b) -> Boolean.compare(
+                        a.getHeight() != HEIGHT_UNKNOWN || a.getWidth() != WIDTH_UNKNOWN,
+                        b.getHeight() != HEIGHT_UNKNOWN || b.getWidth() != WIDTH_UNKNOWN
+                ));
```
```suggestion
                .thenComparing(
                        img -> img.getHeight() != HEIGHT_UNKNOWN || img.getWidth() != WIDTH_UNKNOWN,
                        Boolean::compare
                );
```

Isira-Seneviratne commented on app/src/main/java/org/schabi/newpipe/util/image/ImageStrategy.java:
```diff
@@ -0,0 +1,203 @@
+        // on how close its size is to BEST_LOW_H or BEST_MEDIUM_H (with proper units). Subgroups
+        // without known image size will be left untouched since estimatePixelCount always returns
+        // the same number for those.
+        switch (nonNoneQuality) {
+            case LOW:
+                comparator = comparator.thenComparingDouble(image -> {
+                    final double pixelCount = estimatePixelCount(image, widthOverHeight);
+                    return Math.abs(pixelCount - BEST_LOW_H * BEST_LOW_H * widthOverHeight);
+                });
+                break;
+            default:
+            case MEDIUM:
+                comparator = comparator.thenComparingDouble(image -> {
+                    final double pixelCount = estimatePixelCount(image, widthOverHeight);
+                    return Math.abs(pixelCount - BEST_MEDIUM_H * BEST_MEDIUM_H * widthOverHeight);
+                });
+                break;
+            case HIGH:
+                comparator = comparator.thenComparingDouble(
+                        // this is reversed with a - so that the highest resolution is chosen
+                        i -> -estimatePixelCount(i, widthOverHeight));
+                break;
+        }
```
```suggestion
        final Comparator<Image> finalComparator = switch (nonNoneQuality) {
            case NONE -> initialComparator; // shouldn't be reached
            case LOW -> initialComparator.thenComparingDouble(image -> {
                final double pixelCount = estimatePixelCount(image, widthOverHeight);
                return Math.abs(pixelCount - BEST_LOW_H * BEST_LOW_H * widthOverHeight);
            });
            case MEDIUM -> initialComparator.thenComparingDouble(image -> {
                final double pixelCount = estimatePixelCount(image, widthOverHeight);
                return Math.abs(pixelCount - BEST_MEDIUM_H * BEST_MEDIUM_H * widthOverHeight);
            });
            case HIGH -> initialComparator.thenComparingDouble(
                    // this is reversed with a - so that the highest resolution is chosen
                    i -> -estimatePixelCount(i, widthOverHeight));
        };
```

Isira-Seneviratne commented on app/src/main/java/org/schabi/newpipe/util/image/ImageStrategy.java:
```diff
@@ -0,0 +1,203 @@
+                .orElse(1.0);
+
+        final Image.ResolutionLevel preferredLevel = nonNoneQuality.toResolutionLevel();
+        Comparator<Image> comparator = Comparator
```
```suggestion
        final Comparator<Image> initialComparator = Comparator
```

Isira-Seneviratne commented on app/src/main/java/org/schabi/newpipe/util/image/ImageStrategy.java:
```diff
@@ -0,0 +1,203 @@
+        return images.stream()
+                // using "min" basically means "take the first group, then take the first subgroup,
+                // then choose the best image, while ignoring all other groups and subgroups"
+                .min(comparator)
```
```suggestion
                .min(finalComparator)
```

Isira-Seneviratne reviewed (approved):
This looks good to me overall, I checked on an Android 5.0 emulator and the image loading seems to be working well. I've suggested a few small changes.

JohnBetaro commented:
Hi, I'm here to just make some quick note on your work, it's almost perfect, the image quality has increased a lot compared to before, however it's sharp enough when you compare it to the highest image quality you get when you click on the thumbnail url in the description.
I hope you'll take a look into it, and thank you for your time.
![Screenshot_20230507-230731](https://user-images.githubusercontent.com/114379310/236705110-08f50819-02d5-4a42-950c-bbeb894d2550.png)
![Screenshot_20230507-230809](https://user-images.githubusercontent.com/114379310/236705119-776516f5-39d4-456e-96fe-76b8247112d6.png)
![Screenshot_20230507-231237](https://user-images.githubusercontent.com/114379310/236705122-df24b41d-682a-4df8-81f2-0db686cea36b.png)

at1orianda commented:
Hello, while testing the app CI artifact from commit `cfe4bd39794b67840f576800aef772e2b58f0b13` I encounter this reproducible issue: when manually refreshing the feed the app ui crashes.

Below is the crash log:

## Exception
* User Action: ui error
* Request: ACRA report
* Content Country: US
* Content Language: en-US
* App Language: en_US
* Service: none
* Version: 0.25.1
* OS: Linux Android 12 - 31
<details><summary><b>Crash log </b></summary><p>


java.lang.NullPointerException: Parameter specified as non-null is null: method org.schabi.newpipe.local.feed.service.FeedLoadService$onStartCommand$2.accept, parameter error
 at org.schabi.newpipe.local.feed.service.FeedLoadService$onStartCommand$2.accept(Unknown Source:7)
 at org.schabi.newpipe.local.feed.service.FeedLoadService$onStartCommand$2.accept(FeedLoadService.kt:98)
 at io.reactivex.rxjava3.internal.observers.BiConsumerSingleObserver.onSuccess(BiConsumerSingleObserver.java:56)
 at io.reactivex.rxjava3.internal.operators.single.SingleDoOnSubscribe$DoOnSubscribeSingleObserver.onSuccess(SingleDoOnSubscribe.java:77)
 at io.reactivex.rxjava3.internal.operators.single.SingleObserveOn$ObserveOnSingleObserver.run(SingleObserveOn.java:81)
 at io.reactivex.rxjava3.android.schedulers.HandlerScheduler$ScheduledRunnable.run(HandlerScheduler.java:123)
 at android.os.Handler.handleCallback(Handler.java:938)
 at android.os.Handler.dispatchMessage(Handler.java:99)
 at android.os.Looper.loopOnce(Looper.java:210)
 at android.os.Looper.loop(Looper.java:299)
 at android.app.ActivityThread.main(ActivityThread.java:8273)
 at java.lang.reflect.Method.invoke(Native Method)
 at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:576)
 at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1073)

</details>
<hr>

AudricV commented:
> however it's sharp enough when you compare it to the highest image quality you get when you click on the thumbnail url in the description.

There is nothing we can really do I am afraid.

The extractor returns the thumbnails that YouTube returns on its desktop website for search results. It is not planned to build custom thumbnails ourselves on YouTube outside of RSS feeds and mixes, on which only one thumbnail or no thumbnail is available, as YouTube provides thumbnails without black borders and these thumbnails require signatures and parameters we are not able to generate.

Also, not every thumbnail resolution is available on all videos (look at the extractor PR for more details).

Using different thumbnails would create a different behavior compared to YouTube official clients in more places than RSS feeds and mixes, and I think we like to avoid that as much as we can.

Stypox commented:
- Rebased
- Even though the images extractor PR has been merged, the current NewPipe is incompatible with it because if the channel tab changes, so the NPE commit used at the moment is [this one](https://github.com/Stypox/NewPipeExtractor/commits/pr889-old)
- Added a test for `ImageStrategy` and fixed a boolean condition that should have been the opposite (`image.getHeight() == HEIGHT_UNKNOWN && image.getWidth() == WIDTH_UNKNOWN` instead of the previous `!= || !=`)
- Applied Isira-Seneviratne's suggestions

TobiGr commented on app/src/main/java/org/schabi/newpipe/util/image/PicassoHelper.java:
```diff
@@ -174,19 +183,29 @@ public String key() {
 
-    private static RequestCreator loadImageDefault(final String url, final int placeholderResId) {
+    private static RequestCreator loadImageDefault(@NonNull final List<Image> images,
+                                                   final int placeholderResId) {
```
```suggestion
                                                   @DrawableRes final int placeholderResId) {
```

TobiGr commented on app/src/main/java/org/schabi/newpipe/util/image/PicassoHelper.java:
```diff
@@ -174,19 +183,29 @@ public String key() {
+    }
+
+    private static RequestCreator loadImageDefault(@Nullable final String url,
+                                                   final int placeholderResId) {
```
```suggestion
                                                   @DrawableRes final int placeholderResId) {
```

TobiGr commented on app/src/main/java/org/schabi/newpipe/util/image/PicassoHelper.java:
```diff
@@ -174,19 +183,29 @@ public String key() {
 
-    private static RequestCreator loadImageDefault(final String url, final int placeholderResId,
+    private static RequestCreator loadImageDefault(@Nullable final String url,
+                                                   final int placeholderResId,
```
```suggestion
                                                   @DrawableRes final int placeholderResId,
```

TobiGr reviewed (commented):
LGTM, just some small things

Stypox commented on app/src/main/java/org/schabi/newpipe/util/image/ImageStrategy.java:
```diff
@@ -0,0 +1,195 @@
+
+public final class ImageStrategy {
+
+    // when preferredImageQuality is LOW or MEDIUM, images are sorted by how close their preferred
+    // image quality is to these values (H stands for "Height")
+    private static final int BEST_LOW_H = 75;
+    private static final int BEST_MEDIUM_H = 250;
```
Well, we want this to be a frontend thing, not an extractor thing. We want to be able to change it from NewPipe if we decide to create a better strategy or whatever. Also, these constants represent something different than those in `ResolutionLevel`: the values in `ResolutionLevel` are max-thresholds for the various categories, while these values are used to choose the image whose height is closest to them.

TobiGr replied:
makes sense.

TobiGr replied:
You might need to adjust the values a little, because when setting the image quality to high I get blurry thumbnails for YouTube but perfectly sharp thumbnails for the other services:
<details><summary>Screenshots</summary>

 ![Screenshot_20230921-163515_NewPipe multiple-images](https://github.com/TeamNewPipe/NewPipe/assets/17365767/a2bd50f2-6884-4f52-a8d0-3540e20039b8) ![Screenshot_20230921-163601_NewPipe multiple-images](https://github.com/TeamNewPipe/NewPipe/assets/17365767/08198b3b-1a98-4e94-8101-09cd7ca3339c) ![Screenshot_20230921-163607_NewPipe multiple-images](https://github.com/TeamNewPipe/NewPipe/assets/17365767/59b07908-2957-4fe4-b621-0e7aec65280a) ![Screenshot_20230921-163622_NewPipe multiple-images](https://github.com/TeamNewPipe/NewPipe/assets/17365767/3ff12948-0313-4844-8b76-28763f089b39) 

</details>

Stypox replied:
YouTube doesn't provide better thumbnails for info items unfortunately. The best thumbnail is already being chosen. If you load video details you will see a better picture is loaded for YouTube, too.

TobiGr replied:
ah ok. In this case we'll need to do some changes in the extractor. I just checked the thumbnail URLs. They look like this:

Stream URL: https://www.youtube.com/watch?v=M9k_2aX7Z24
https://i.ytimg.com/vi/M9k_2aX7Z24/hqdefault.jpg?sqp=-oaymwEbCMQBEG5IVfKriqkDDggBFQAAiEIYAXABwAEG&rs=AOn4CLDkVXxfTq7DmG9_QfZ9cZtHa33QNQ
https://i.ytimg.com/vi/M9k_2aX7Z24/hqdefault.jpg?sqp=-oaymwEcCPYBEIoBSFXyq4qpAw4IARUAAIhCGAFwAcABBg==&rs=AOn4CLBIEhn3OK2WJ02HqBfzaYxw_2KrYw
https://i.ytimg.com/vi/M9k_2aX7Z24/hqdefault.jpg?sqp=-oaymwEcCNACELwBSFXyq4qpAw4IARUAAIhCGAFwAcABBg==&rs=AOn4CLCj_e440zgW4wR6f5qyZl8KikrUZA
https://i.ytimg.com/vi/M9k_2aX7Z24/maxresdefault.jpg

Stream URL: https://www.youtube.com/watch?v=b3unvqHR9m8
https://i.ytimg.com/vi/b3unvqHR9m8/hqdefault.jpg?sqp=-oaymwEcCPYBEIoBSFXyq4qpAw4IARUAAIhCGAFwAcABBg==&rs=AOn4CLCci958CBvtcsnyOoXY23t-Jngdeg
https://i.ytimg.com/vi/b3unvqHR9m8/hqdefault.jpg?sqp=-oaymwEcCNACELwBSFXyq4qpAw4IARUAAIhCGAFwAcABBg==&rs=AOn4CLC0KAPC8whJ6NYDrgFURSLezjLHQQ
https://i.ytimg.com/vi_webp/b3unvqHR9m8/maxresdefault.webp?v=650ab64d

Although the last URL is different from the maxres URL from above, this one also works, but is not provided by YouTube:
https://i.ytimg.com/vi/b3unvqHR9m8/maxresdefault.jpg

We might be able to just add them in the extractor even if the highres version is not provided. The URL seems to be
`https://i.ytimg.com/vi/<STREAM_ID>/maxresdefault.jpg`

CC @AudricV

Stypox replied:
I think this can be taken care of separately simply with an update in the extractor

AudricV replied:
As we discussed on our internal chat, such thumbnails are not provided by YouTube desktop client and `maxresdefault` variant is not available on some videos. This behavior will not change.

Stypox commented on app/src/main/java/org/schabi/newpipe/fragments/detail/DescriptionFragment.java:
```diff
@@ -187,6 +201,74 @@ private void addMetadataItem(final LayoutInflater inflater,
 
+    private String imageSizeToText(final int heightOrWidth) {
+        if (heightOrWidth < 0) {
+            return "?";
```
Thanks for noticing, I created a `question_mark` string.
